### PR TITLE
Use mlx5 vf as PF1 proxy

### PR DESCRIPTION
**Pull request by PlagueCZ**
Given TAP device's limitations, we implemented the PF1-proxy using a VF bound to MLX5 driver (as opposed to VFIO). This VF is created on PF1 to make it easier to set up.

This also enables offloading of the proxy path. For normal operation this is easy, for OSC vith virtual services, there is a little bit more work, but nothing major.

I tried to reorder and squash commits into blocks of work, so this PR should be nicely readable by going through it commit-by-commit.

I had to change the prepare script, but it should be backwards-compatible. There is one reordering (setting switchdev before setting the number of VFs) that is needed for multiport mode. OSC is also now provisioning the HW differently, so I only tested the prepare script's HW setup by hand locally. 

Connected to #606

### CODE REVIEW

File: include/rte_flow/dp_rte_flow_helpers.h [hunk 1]
```diff
@@ -62,6 +68,18 @@ static const struct rte_flow_item_ipv6 dp_flow_item_ipv6_dst_mask = {
 	.hdr.dst_addr = "\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff",
 	.hdr.proto = 0xff,
 };
+#ifdef ENABLE_VIRTSVC
+static const struct rte_flow_item_ipv6 dp_flow_item_ipv6_src_dst_mask = {
```
[the remark this answers is not in the document]
Comment (PlagueCZ) [reply include/rte_flow/dp_rte_flow_helpers.h:72]: Well they all contain proto (dp_flow_item_ipv6_mask, dp_flow_item_ipv6_src_mask, dp_flow_item_ipv6_dst_mask), that's why I kept it and the one without proto is dp_flow_item_ipv6_dst_only_mask

But we can rename all of these if you like.

File: include/dp_port.h [hunk 2]
```diff
@@ -57,6 +57,10 @@ struct dp_port_async_template {
 
 enum dp_port_async_template_type {
 	DP_PORT_ASYNC_TEMPLATE_PF_ISOLATION,
+#ifdef ENABLE_PF1_PROXY
+	DP_PORT_ASYNC_TEMPLATE_PF1_FROM_PROXY,
```
[the remark this answers is not in the document]
Comment (PlagueCZ) [reply include/dp_port.h:61]: The idea was that "pf1 to proxy" and "proxy to pf1" seem similar, whereas "from" is directly telling me the direction is reversed. But if you prefer I can simply rename.

Though the some more functions would need renaming, like `dp_create_pf_async_from_proxy_templates`, `dp_create_pf_async_to_proxy_templates` etc.

File: src/rte_flow/dp_rte_async_flow_isolation.c [hunk 3]
```diff
@@ -220,31 +224,36 @@ int dp_create_pf_async_isolation_rules(struct dp_port *port)
...
 	} else {
 		port->default_async_rules.default_flows[DP_PORT_ASYNC_FLOW_ISOLATE_IPV6] = flow;
 		rule_count++;
 	}
 
+#ifdef ENABLE_PF1_PROXY
+	if (dp_conf_is_pf1_proxy_enabled() && port == dp_get_pf0()) {
+		rules_required += DP_PF1_PROXY_RULE_COUNT;
+		rule_count += dp_create_pf1_proxy_async_isolation_rules(port);
+		// cannot end on error, need to commit partial success
```
[the remark this answers is not in the document]
Comment (PlagueCZ) [reply src/rte_flow/dp_rte_async_flow_isolation.c:238]: We would no like to keep going as this impacts performance and we found that Ceph storage degrades rapidly making the stack unusable, so failing is preferred.

Setting the link down would then "disable" the possibility of updating the switch, which is also not something we can allow.

**Comment by PlagueCZ** [comment]
As for the graph schema, I thought about it, but given that virtual services are also not there because they are in an `ifdef`, I dod not do it.

But if it's wanted I could create two copies of the shema and create a virtual-services-specific one and pf1-proxy-specific one.

File: src/rte_flow/dp_rte_async_flow_isolation.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (PlagueCZ) [reply src/rte_flow/dp_rte_async_flow_isolation.c]: Updated comments to be more clear

File: src/dp_port.c [hunk 4]
```diff
@@ -581,16 +553,26 @@ static int dp_port_install_async_isolated_mode(struct dp_port *port)
...
 {
-	DPS_LOG_INFO("Installing PF async templates", DP_LOG_PORTID(port->port_id));
+	DPS_LOG_INFO("Installing PF async templates", DP_LOG_PORT(port));
 	if (DP_FAILED(dp_create_pf_async_isolation_templates(port))) {
-		DPS_LOG_ERR("Failed to create pf async isolation templates", DP_LOG_PORTID(port->port_id));
+		DPS_LOG_ERR("Failed to create pf async isolation templates", DP_LOG_PORT(port));
 		return DP_ERROR;
 	}
+#ifdef ENABLE_PF1_PROXY
+	if (dp_conf_is_pf1_proxy_enabled() && port == dp_get_pf0()) {
```
[the remark this answers is not in the document]
Comment (PlagueCZ) [reply src/dp_port.c:563]: Good point, added a comment

**Review by byteocean — APPROVED** [review]
LGTM.

**Comment by guvenc** [comment]
@PlagueCZ 
I am still going through the changes but one thing which popped into my mind is that clang build which is run anyways and produces throw away binary during the CI/CD can include the compiler flag of pf1 proxy so that this newly introduced code path gets compiled at least once during CI/CD.

**Comment by PlagueCZ** [comment]
We should really run a tester in a job in OSC, but unfortunately that is still not being done.
We should then add another compilation run though, because I already had clang catch a warning about unused variable, which was unused due to pf1-proxy *not* being there

**Review by guvenc — CHANGES_REQUESTED** [review]
@PlagueCZ 
I think this PR is in good shape. Thanks also for providing the updated dataplane graph schema.
- Let's add a (temporary) compilation step in Dockerfile where this newly introduced feature is enabled, as discussed.

- And also the clarification whether the RoCE needs to be disabled on firmware level to get a good performance when mpesw is enabled. If so, documenting it in the docs with a sentence or so.

- And also please rebase to main. 

Apart from these things, in pretty good shape. Thanks !

**Review by guvenc — APPROVED** [review]

**Pull request merged**